File: src/main/java/pl/allegro/experiments/chi/chiserver/infrastructure/experiments/MongoExperimentsRepository.java
package pl.allegro.experiments.chi.chiserver.infrastructure.experiments;

import io.micrometer.core.instrument.Timer;
import org.apache.commons.collections.IteratorUtils;
import org.javers.core.Javers;
import org.springframework.data.mongodb.core.MongoTemplate;
import org.springframework.stereotype.Repository;
import pl.allegro.experiments.chi.chiserver.domain.UserProvider;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentDefinition;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;
import java.util.stream.Collectors;

@Repository
class MongoExperimentsRepository implements ExperimentsRepository {
    private final ExperimentsMongoMetricsReporter experimentsMongoMetricsReporter;
    private final Javers javers;
    private final UserProvider userProvider;
    private final ExperimentDefinitionCrudRepository experimentDefinitionCrudRepository;

    public MongoExperimentsRepository(ExperimentsMongoMetricsReporter experimentsMongoMetricsReporter, Javers javers, UserProvider userProvider, ExperimentDefinitionCrudRepository experimentDefinitionCrudRepository) {
        this.experimentsMongoMetricsReporter = experimentsMongoMetricsReporter;
        this.javers = javers;
        this.userProvider = userProvider;
        this.experimentDefinitionCrudRepository = experimentDefinitionCrudRepository;
    }

    @Override
    public Optional<ExperimentDefinition> getExperiment(String id) {
        Timer timer = experimentsMongoMetricsReporter.timerSingleExperiment();
        Optional<ExperimentDefinition> experiment = timer.record(
            () -> experimentDefinitionCrudRepository.findById(id));
        return experiment;
    }

    @Override
    public void delete(String experimentId) {
        ExperimentDefinition definition = getExperiment(experimentId).orElse(null);
        String username = userProvider.getCurrentUser().getName();
        javers.getLatestSnapshot(experimentId, ExperimentDefinition.class).ifPresent(it ->
            javers.commitShallowDelete(username, definition)
        );
        if (definition != null) {
            experimentDefinitionCrudRepository.delete(definition);
        }
    }

    @Override
    public void save(ExperimentDefinition experimentDefinition) {
        String username = userProvider.getCurrentUser().getName();
        javers.commit(username, experimentDefinition);
        experimentDefinitionCrudRepository.save(experimentDefinition);
    }

    @Override
    public List<ExperimentDefinition> getAll() {
        Timer timer = experimentsMongoMetricsReporter.timerAllExperiments();
        return timer.record(() -> IteratorUtils.toList(experimentDefinitionCrudRepository.findAll().iterator()));
    }

    @Override
    public List<ExperimentDefinition> assignable() {
        return getAll().stream()
                .filter(ExperimentDefinition::isAssignable)
                .collect(Collectors.toList());
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/infrastructure/experiments/MongoExperimentsRepository.kt
package pl.allegro.experiments.chi.chiserver.infrastructure.experiments

import org.javers.core.Javers
import org.springframework.data.mongodb.core.MongoTemplate
import pl.allegro.experiments.chi.chiserver.domain.UserProvider
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository

private const val COLLECTION = "experiments"

open class MongoExperimentsRepository(val mongoTemplate: MongoTemplate,
                                      val experimentsMongoMetricsReporter: ExperimentsMongoMetricsReporter,
                                      val javers: Javers,
                                      val userProvider: UserProvider) :
        ExperimentsRepository {
    override fun getExperiment(id: String): Experiment? {
        experimentsMongoMetricsReporter.timerSingleExperiment().use {
            return mongoTemplate.findById(id, Experiment::class.java, COLLECTION)
        }
    }

    override fun delete(experimentId: String) {
        val experiment = getExperiment(experimentId)
        val userName = userProvider.currentUser.name
        javers.getLatestSnapshot(experimentId, Experiment::class.java).ifPresent {
            javers.commitShallowDelete(userName, experiment)
        }
        mongoTemplate.remove(experiment, COLLECTION)
    }

    override fun save(experiment: Experiment) {
        val userName = userProvider.currentUser.name
        javers.commit(userName, experiment)
        mongoTemplate.save(experiment, COLLECTION)
    }

    override fun getAll(): List<Experiment> {
        experimentsMongoMetricsReporter.timerAllExperiments().use {
            return mongoTemplate.findAll(Experiment::class.java, COLLECTION)
        }
    }

    override fun assignable(): List<Experiment> {
        return getAll().filter { it.isAssignable() }
    }
}